### PR TITLE
UGNI: Support EP flush with callbacks

### PR DESCRIPTION
**Pull request by MattBBaker**
Replace terrible previous hack with new right and proper implementation.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1545/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3464/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe can you please take a look as well.

**Comment by shamisp** [comment]
@yosefe please take a look

File: src/uct/ugni/base/ugni_ep.c [hunk 1]
```diff
@@ -88,6 +78,84 @@ void uct_ugni_ep_pending_purge(uct_ep_h tl_ep,
...
+
+    if (!uct_ugni_ep_can_send(ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    if (NULL == comp) {
+        return UCS_INPROGRESS;
+    }
+
+    new_group = uct_ugni_new_flush_group(iface);
```
Comment (manjugv) [inline src/uct/ugni/base/ugni_ep.c]: @MattBBaker I think you might to protect here for multithreaded case.

Comment (MattBBaker) [inline src/uct/ugni/base/ugni_ep.c]: Well, protect in the get function itself. I may also look into lockless mpools.

**Pull request merged**